core/market/decentralized/src/protocol/discovery.rs

```diff
@@ -34,21 +46,182 @@ pub enum DiscoveryInitError {
+    Ok(())
 }
 
-/// Creates Discovery of specific type.
-pub trait DiscoveryFactory {
-    fn new(builder: DiscoveryBuilder) -> Result<Arc<dyn Discovery>, DiscoveryInitError>;
+async fn retrieve_offers(myself: Arc<DiscoveryImpl>) -> Result<Vec<Offer>, DiscoveryError> {
+    unimplemented!()
 }
 
+async fn on_offer_received(
+    myself: Arc<DiscoveryImpl>,
+    caller: String,
+    msg: OfferReceived,
+) -> Result<(), ()> {
+    let callback = myself.offer_received.clone();
+
+    let offer = msg.offer.clone();
+    let offer_id = offer.offer_id().unwrap_or("{Empty id}").to_string();
+    let provider_id = offer.provider_id().unwrap_or("{Empty id}").to_string();
+
+    log::info!(
+        "Received broadcasted Offer [{}] from provider [{}]. Sender: [{}].",
+        offer_id,
+        provider_id,
+        &caller,
+    );
+
+    match callback.call(caller, msg).await? {
+        PropagateOffer::True => {
+            log::info!("Propagating further Offer [{}].", offer_id,);
+
+            // TODO: Should we retry in case of fail?
+            if let Err(error) = broadcast_offer(myself, offer).await {
+                log::error!(
+                    "Error propagating further Offer [{}] from provider [{}].",
+                    offer_id,
+                    provider_id
+                );
+            }
+        }
+        PropagateOffer::False(reason) => {
+            log::info!(
+                "Not propagating Offer [{}] for reason: {}.",
+                offer_id,
+                reason
+            );
+        }
+    }
+    Ok(())
+}
+
+// =========================================== //
+// Discovery messages
+// =========================================== //
+
+#[derive(Serialize, Deserialize, Display)]
+pub enum StopPropagateReason {
+    #[display(fmt = "Offer already exists in database")]
+    AlreadyExists,
+    #[display(fmt = "Error adding offer: {}", "_0")]
+    Error(String),
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum PropagateOffer {
+    True,
+    False(StopPropagateReason),
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OfferReceived {
+    pub offer: Offer,
+}
+
+impl RpcMessage for OfferReceived {
+    const ID: &'static str = "OfferReceived";
+    type Item = PropagateOffer;
+    type Error = ();
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct RetrieveOffers {
+    pub newer_than: chrono::DateTime<Utc>,
+}
+
+impl RpcMessage for RetrieveOffers {
+    const ID: &'static str = "RetrieveOffers";
+    type Item = Vec<Offer>;
+    type Error = DiscoveryRemoteError;
+}
+
+// =========================================== //
+// Internal Discovery messages used
+// for communication between market instances
+// of Discovery protocol
+// =========================================== //
+
+impl BroadcastMessage for OfferReceived {
+    const TOPIC: &'static str = "market-protocol-mk1-offer";
```

Comment:
maybe sth shorter like
```suggestion
    const TOPIC: &'static str = "new-offer";
```

Reply:
This should contain at least market-mk1 prefix, to avoid clashes with other Services that can potentially register it's endpoints